### PR TITLE
Vamas sub-reader: enhance data extraction, add IRREGULAR parsing

### PR DESCRIPTION
**Pull request by lukaspie**
This enhances the capabilities of the existing Vamas XPS reader.
The Vamas reader reads .vms files, which is the current ISO standard data transfer format ([ISO 14976](https://www.iso.org/standard/24269.html)).

VMS specific:
- [x] Write CPS data to file
- [x]  Extract units for quantities
- [x] Parse header comments -> additional metadata (depends on spectrometer software that writes the vms file)
- [x] Parse comment lines -> data analysis
- [x] Parse vamas files of type IRREGULAR
- [x] Update config file (will be done again when we have the [*] syntax from the MultiFormatReader

General:
- [x] align mapper dictionary keys

### CODE REVIEW

File: pynxtools_xps/vms/vamas.py [hunk 1]
```diff
@@ -162,23 +180,47 @@ def _update_xps_dict_with_spectrum(self, spectrum, key_map):
...
+        # Write process data
+        process_key_map: Dict[str, List[str]] = {
+            "energy_referencing": ["alignments"],
+            "peak_fitting": ["regions", "components"],
+        }
+
+        for grouping, process_key_list in process_key_map.items():
+            root = path_map[str(grouping)]
+            for spectrum_key in process_key_list:
+                try:
```
Comment (RubelMozumder) [inline pynxtools_xps/vms/vamas.py:212]: No need to convert into `str` as you already maping `process_key_map` from `str` to `list`. is needed can be check is `str` or other types.

**Comment by RubelMozumder** [comment]
Here are the some change requests. 
I see this is a big code base. So, you can add some tests specific to the reader according to the file extension.

File: pynxtools_xps/vms/vamas.py (hunk 1 above)
Comment (lukaspie) [reply pynxtools_xps/vms/vamas.py:212]: removed the string conversion.

**Comment by lukaspie** [comment]
> Here are the some change requests. I see this is a big code base. So, you can add some tests specific to the reader according to the file extension.

I have added test data for the Vamas format that is addressed by this PR. Data for other file extensions will be added later.

File: pynxtools_xps/vms/vamas.py [hunk 2]
```diff
@@ -46,14 +52,27 @@ class VamasMapper(XPSMapper):
...
     def __init__(self):
-        self.file = None
-        self.parsers = [
-            VamasParserRegular,
-            VamasParserIrregular,
-        ]
-        self.parser_map = {
-            "regular": VamasParserRegular,
-            "irregular": VamasParserIrregular,
+        self.file: Union[str, Path] = ""
```
[the remark this answers is not in the document]
Comment (lukaspie) [reply pynxtools_xps/vms/vamas.py]: Btw, I moved this code to the abstract base class, but it still applies. https://github.com/FAIRmat-NFDI/pynxtools-xps/blob/acf658137e148a329ac8c065112d649a4a99170d/pynxtools_xps/reader_utils.py#L55

File: pynxtools_xps/vms/vamas.py [hunk 3]
```diff
@@ -46,14 +52,26 @@ class VamasMapper(XPSMapper):
...
-        self.parsers = [
-            VamasParserRegular,
-            VamasParserIrregular,
-        ]
-        self.parser_map = {
-            "regular": VamasParserRegular,
-            "irregular": VamasParserIrregular,
+        self.parsers: List[Any] = []
+
+        self.units: dict = {
```
Comment (domna) [inline pynxtools_xps/vms/vamas.py:57]: This could also be a top-level constant, no? Or do you add something while parsing?

Comment (lukaspie) [reply pynxtools_xps/vms/vamas.py:57]: I moved it out, thanks for the idea.

**Comment by lukaspie** [comment]
> One small comment on the test: You could parametrise it with the reader directories. That way you would get a single run for each specific reader and would immediately see which of them fails. It's maybe something we could also introduce in the plugin template

Good idea. I think I will address this in another PR since I want to add some more tests for the individual readers anyway.

File: pynxtools_xps/vms/vamas.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply pynxtools_xps/vms/vamas.py]: Just for clearification. I meant file could be a complete path of the file or a file without a full path. But `str` obj can be read by os lib and Path can be read by `pathlib`. However if you add the functionality for both of them that is fine too.

**Review by RubelMozumder — CHANGES_REQUESTED** [review]

File: pynxtools_xps/vms/vamas.py [hunk 4]
```diff
@@ -659,30 +577,44 @@ def _parse_norm_block(self):
...
+        if self.header.scan_mode == "REGULAR":
+            block.abscissa_label = self.data.pop(0).strip()
+            block.abscissa_units = self.data.pop(0).strip()
+            block.abscissa_start = float(self.data.pop(0).strip())
+            block.abscissa_step = float(self.data.pop(0).strip())
+
+            block.no_variables = int(self.data.pop(0).strip())
+            for var in range(block.no_variables):
+                name = "variable_label_" + str(var + 1)
+                setattr(block, name, self.data.pop(0).strip())
```
[the remark this answers is not in the document]
Comment (lukaspie) [reply pynxtools_xps/vms/vamas.py:598]: No, this doesn't work. There are no keywords in the vms file and the individual elements HAVE to be there in order for it to conform to the VAMAS standard. basically, the order of elements is pre-fixed. So there is no other option than parsing the data as a list and work with the `list.pop` function.

**Comment by RubelMozumder** [comment]
Sorry for the confusion. I reviewed some code on the first review but some of them submitted and some of them I did not submit. So, now you see new change requests. Which I am going to resolve now. I think there a few comment you may add in the future especially [pynxtools_xps/vms/vamas.py](https://github.com/FAIRmat-NFDI/pynxtools-xps/pull/10/files/e1cbf32d0fd6fd212ad5ddb0b4f490dc66cd840f#diff-82024b209bae26337d81594545a33b118bbf0ea0ad82a2ab928d11e85154c7b9) where you assumed all the data will come in a chain. But if one of them is broken or missing other fields will read the wrong data. So in another PR you can take care of it.

File: pynxtools_xps/vms/vamas.py [hunk 5]
```diff
@@ -828,208 +788,300 @@ def _add_data_values(self, block):
...
             setattr(block, name, data_dict[name])
 
+    def _add_irregular_data(self, block: VamasBlock):
+        """Parse data with regularly spaced energy axis."""
+        data_dict: Dict[str, List] = {}
 
-# THIS DOESN'T WORK SO FAR!!
-class VamasParserIrregular(VamasParser):
-    """Parser for .vms files of type IRREGULAR"""
+        block_data = list(np.array(self.data[: block.num_ord_values], dtype=float))
```
[the remark this answers is not in the document]
Comment (lukaspie) [reply pynxtools_xps/vms/vamas.py:789]: I removed the conversion to a list.

File: pynxtools_xps/vms/vamas.py [hunk 6]
```diff
@@ -133,26 +144,33 @@ def construct_data(self):
...
-                "scans",
             ],
         }
 
         for spectrum in spectra:
             self._update_xps_dict_with_spectrum(spectrum, key_map)
 
-    def _update_xps_dict_with_spectrum(self, spectrum, key_map):
+    def _update_xps_dict_with_spectrum(
+        self, spectrum: Dict[str, Any], key_map: Dict[str, str]
```
[the remark this answers is not in the document]
Comment (lukaspie) [reply pynxtools_xps/vms/vamas.py:158]: done

File: pynxtools_xps/vms/vamas.py [hunk 7]
```diff
@@ -828,208 +788,300 @@ def _add_data_values(self, block):
...
+
+        special_keys = {
+            "Casa Info Follows": self._handle_casa_header,
+            "SpecsLab Prodigy": self._handle_prodigy_header,
+            # "SOFH": self._handle_phi_header,
+        }
+
+        for keyword, handle_func in special_keys.items():
+            if any(keyword in line for line in comment_list):
+                index = [i for i, line in enumerate(comment_list) if keyword in line][0]
```
[the remark this answers is not in the document]
Comment (lukaspie) [reply pynxtools_xps/vms/vamas.py:865]: Changed 👍

**Comment by lukaspie** [comment]
> Sorry for the confusion. I reviewed some code on the first review but some of them submitted and some of them I did not submit. So, now you see new change requests. Which I am going to resolve now. I think there a few comment you may add in the future especially [pynxtools_xps/vms/vamas.py](https://github.com/FAIRmat-NFDI/pynxtools-xps/pull/10/files/e1cbf32d0fd6fd212ad5ddb0b4f490dc66cd840f#diff-82024b209bae26337d81594545a33b118bbf0ea0ad82a2ab928d11e85154c7b9) where you assumed all the data will come in a chain. But if one of them is broken or missing other fields will read the wrong data. So in another PR you can take care of it.

There is really no other way around then reading the data in a list. The Vamas format does not allow for keywords in the file and it alway assumes that all of the elements are at the correct positions (it is an ISO standard afterall, so if you don't have the right structure, it cannot be a vamas file).

**Pull request merged**